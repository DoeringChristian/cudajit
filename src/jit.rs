use std::borrow::BorrowMut;
use std::collections::{HashMap, HashSet};
use std::fmt::Write;

use once_cell::sync::Lazy;
use parking_lot::Mutex;

use crate::backend::Kernel;
use crate::schedule::ScheduleIr;
use crate::trace::Internal;
use crate::var::{Op, VarId};

// TODO: pooling for paralel exectution
///
/// The Jit Compiler first generates schedules (Intermediate Representation) from a Trace.
/// It then assembles and compiles a Kernel depending on the Backend.
///
/// Ir -> [Schedule; N] -> [Kernel; N]
///
/// Where N is the number of schedule groups.
/// These are extracted from the scheduled variables in the Ir.
///
#[derive(Debug, Default)]
pub struct Jit {
    pub schedules: Vec<ScheduleIr>,
    pub kernels: Vec<Box<dyn Kernel>>,
}

///
/// This represents the Evaluation of one or many variables.
/// With dependencies on other variables which have been scheduled.
/// TODO: Better naming
///
/// Dependencies of a scatter operation have to be set up in the following way:
///  a  <---------- b
/// ^ ^------ s1 <--|
/// '-- s0 <--|
///
/// A is an arbitrary operation. If s0 and s1 are scattering operations, a has to be scheduled as
/// do s0 and s1. In order for s0 and s1 to be evaluated before b, b needs to depend on s1 and s1
/// on s0.
/// This can be done by adding extra dependencies to b and s1.
///
///
#[derive(Debug)]
struct Pass {
    ids: Vec<VarId>,
    deps: Vec<usize>,
    size: usize,
}
///
/// Does a depend on b?
///
fn depends_on(passes: &[Pass], a: usize, b: usize) -> bool {
    if a == b {
        return true;
    }
    for dep in passes[b].deps.iter() {
        if depends_on(passes, a, *dep) {
            return true;
        }
    }
    return false;
}
fn try_merge(passes: &mut Vec<Pass>, dst: usize, src: usize) -> Option<()> {
    if passes[dst].size != passes[src].size {
        return None;
    }
    if depends_on(&passes, dst, src) {
        return None;
    }

    let src_pass = passes.remove(src);

    for pass in passes[src..].iter_mut() {
        for dep in pass.deps.iter_mut() {
            if *dep == src {
                *dep = dst;
            } else if *dep > src {
                *dep -= 1;
            }
        }
    }

    passes[dst].deps.extend_from_slice(&src_pass.deps);
    passes[dst].ids.extend_from_slice(&src_pass.ids);

    Some(())
}

impl Jit {
    ///
    /// Evaluates a Ir by first constructing Schedules, which are then compiled and assembled
    /// into kernels.
    ///
    /// A the end, all scheduled variables are overwritten with the calculated data.
    ///
    pub fn eval(&mut self, ir: &mut Internal) {
        // For every scheduled variable (destination) we have to create a new buffer (except if it
        // is void)
        for id in ir.scheduled.clone() {
            let var = ir.var(id);
            // Do not reallocate on scheduled variables (don't yet know if this is right)
            if var.buffer.is_none() && var.ty.size() > 0 {
                let size = ir.var(id).size;
                let ty_size = ir.var(id).ty.size();
                let buffer = ir.backend.as_ref().unwrap().buffer_uninit(size * ty_size);

                let mut var = ir.var_mut(id);
                var.buffer = Some(buffer);
            }
        }

        self.compile(ir);
        let n_kernels = self.kernels.len();
        for i in 0..n_kernels {
            let (kernel, schedule) = self.schedule_kernel(i);
            kernel.execute_async(schedule);
            ir.backend.as_ref().unwrap().synchronize(); // TODO: only sync if we have
                                                        // dependencies
        }

        // After executing the kernels, the Ir is cleaned up.
        // To do so, we first decrement the refcount and then set the ParamType to Input and op to
        // Data
        for id in ir.scheduled.clone() {
            let var = ir.var_mut(id);

            // Set op and type for next kernel:
            // var.param_ty = ParamType::Input;
            var.op = Op::Data;

            // Clear dependencies:
            let deps = var.deps.clone();
            var.deps.clear();

            for dep in deps {
                ir.dec_rc(dep);
            }

            ir.dec_rc(id);
        }
        ir.scheduled.clear();
    }
    ///
    /// Collect neccesary passes from `ir`.
    /// NOTE: We would need to correctly assign dependencies on scatter
    /// NOTE: Passes are in ordered (DAG ordering)
    ///
    fn passes(&self, ir: &Internal) -> Vec<Pass> {
        ///
        /// Gets the dependencies of `id` in `scheduled`
        ///
        fn dependencies_in(
            ir: &Internal,
            id: VarId,
            scheduled: &HashSet<VarId>,
            deps: &mut Vec<VarId>,
        ) {
            if scheduled.contains(&id) {
                deps.push(id);
            }
            let var = ir.var(id);
            for dep in var.deps.iter() {
                dependencies_in(ir, *dep, scheduled, deps);
            }
        }
        let scheduled = ir.scheduled.iter().cloned().collect::<HashSet<_>>();
        let mut id2pass = HashMap::new();
        let passes = ir
            .scheduled
            .iter()
            .enumerate()
            .map(|(i, id)| {
                let mut deps = Vec::new();
                dependencies_in(ir, *id, &scheduled, &mut deps);
                let deps = deps
                    .into_iter()
                    .filter(|dep| dep != id)
                    .map(|dep| id2pass[&dep])
                    .collect();
                id2pass.insert(*id, i);
                Pass {
                    ids: vec![*id],
                    deps,
                    size: ir.var(*id).size,
                }
            })
            .collect::<Vec<_>>();

        // TODO: use passes to flatten dependencies and stop traversal eraly.

        passes
    }
    ///
    /// Compiles the computation graph of all scheduled variables in a Ir.
    ///
    /// First, all scheduled variables with the same size are grouped.
    /// Then, a Schedule Intermediate Representation is constructed from the groups.
    /// In the end a set of Kernels is assembled and compiled.
    ///
    fn compile(&mut self, ir: &Internal) {
        if ir.scheduled.len() == 0 {
            return;
        }
        self.schedules.clear();
        self.kernels.clear();

        // The problem we try to solve is the following:
        //
        // The nodes in `ir` are represent DAG, so do the `passes`.
        // They are also stored in a topological ordering.
        // We try to merge as many passes as possible.
        // Passes can only be merged if tey have the same size.
        //
        // This algorithm is not perfect and results in a potentially sub optimal result.
        // Also, using hash sets might not be the best approach.
        let mut passes = self.passes(&ir);

        for i in (0..passes.len()).rev() {
            for j in (0..i).rev() {
                // Try to merge i into j
                if try_merge(&mut passes, j, i).is_some() {
                    break;
                }
            }
        }
        let first_register = ir.backend.as_ref().unwrap().first_register();
        self.schedules = passes
            .iter()
            .map(|pass| {
                let mut s = ScheduleIr::new(first_register, pass.size);
                s.collect_vars(ir, &pass.ids.iter().cloned().collect::<Vec<_>>());
                s
            })
            .collect::<Vec<_>>();

        // TODO: paralelize by populating kernels first.
        self.kernels = self
            .schedules
            .iter()
            .map(|mut s| {
                let mut kernel = ir.backend.as_ref().unwrap().new_kernel();
                kernel.assemble(&mut s);
                // kernel.compile();
                kernel
            })
            .collect::<Vec<_>>();
        for kernel in self.kernels.iter_mut() {
            kernel.compile();
        }
    }
    pub fn schedule_kernel(&mut self, i: usize) -> (&mut Box<dyn Kernel>, &mut ScheduleIr) {
        (&mut self.kernels[i], &mut self.schedules[i])
    }
    ///
    /// Writes the kernel assemblies into a string which can then be checked by snapshot testing
    /// tools such as insta.
    ///
    pub fn kernel_debug(&self) -> String {
        let mut string = String::new();
        for (i, k) in self.kernels.iter().enumerate() {
            writeln!(string, "===============================================").unwrap();
            writeln!(string, "Kernel {}:", i).unwrap();
            writeln!(string, "").unwrap();
            write!(string, "{}", k.assembly()).unwrap();
        }
        string
    }
}

#[cfg(test)]
mod test {}
