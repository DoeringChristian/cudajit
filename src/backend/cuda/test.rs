use crate::jit::Jit;
use crate::trace::{ReduceOp, Trace};

#[test]
fn refcounting() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.buffer_f32(&[1.; 10]);
    assert_eq!(x.var().rc, 1, "rc of x should be 1 (in x)");
    let y = x.add(&x);
    // let y = ir::add(&x, &x);

    assert_eq!(
        x.var().rc,
        3,
        "rc of x should be 3 (2 refs in y and 1 in x)"
    );

    assert_eq!(y.var().rc, 1, "rc of y should be 1 (in y)");

    ir.schedule(&[&y]);
    let mut jit = Jit::default();

    assert_eq!(
        y.var().rc,
        2,
        "rc of y should be 2 (1 in y and 1 in schedule)"
    );

    jit.eval(&mut ir.borrow_mut());

    assert_eq!(
        x.var().rc,
        1,
        "rc of x should be 1 (dependencies of y shuld be cleaned)"
    );
    assert_eq!(
        y.var().rc,
        1,
        "rc of y should be 2 (y from schedule should be cleaned)"
    );

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(y.to_host_f32(), vec![2f32; 10]);
}
#[test]
fn load_add_f32() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.buffer_f32(&[1.; 10]);
    // let y = ir::add(&x, &x);
    let y = x.add(&x);

    ir.schedule(&[&y]);
    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(y.to_host_f32(), vec![2f32; 10]);
}
#[test]
fn load_gather_f32() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.buffer_f32(&[1., 2., 3., 4., 5.]);
    let i = ir.buffer_u32(&[0, 1, 4]);
    let y = x.gather(&i, None);

    ir.schedule(&[&y]);
    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(y.to_host_f32(), vec![1., 2., 5.]);
}
#[test]
fn reindex() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.index(10);

    let i = ir.index(3);
    let c = ir.literal_u32(2);
    let i = i.add(&c);

    let y = x.gather(&i, None);

    ir.schedule(&[&y]);
    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(y.to_host_u32(), vec![2, 3, 4]);
}
#[test]
fn index() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let i = ir.index(10);

    ir.schedule(&[&i]);
    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(i.to_host_u32(), vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
}
#[test]
fn gather_eval() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let tmp_x;
    let tmp_y;
    let tmp_z;
    let r = {
        let x = ir.index(3);
        dbg!();
        assert_eq!(x.var().rc, 1);
        let y = ir.buffer_u32(&[1, 2, 3]);
        dbg!();
        assert_eq!(y.var().rc, 1);

        // let z = ir::add(&x, &y);
        let z = x.add(&y);
        dbg!();
        assert_eq!(x.var().rc, 2);
        assert_eq!(y.var().rc, 2);
        assert_eq!(z.var().rc, 1);

        let r = z.gather(&ir.index(3), None);
        dbg!();
        assert_eq!(x.var().rc, 2);
        assert_eq!(y.var().rc, 2);
        assert_eq!(z.var().rc, 3);
        assert_eq!(r.var().rc, 1);
        tmp_x = x.id();
        tmp_y = y.id();
        tmp_z = z.id();
        r
    };
    assert_eq!(r.var().rc, 1);
    assert_eq!(ir.borrow_mut().get_var(tmp_x).unwrap().rc, 1);
    assert_eq!(ir.borrow_mut().get_var(tmp_y).unwrap().rc, 1);
    assert_eq!(ir.borrow_mut().get_var(tmp_z).unwrap().rc, 2); // z is referenced by r and the
                                                               // schedule

    ir.schedule(&[&r]);
    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    assert_eq!(r.var().rc, 1);
    assert!(ir.borrow_mut().get_var(tmp_z).is_none());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(r.to_host_u32(), vec![1, 3, 5]);
}
#[test]
fn paralell() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.index(10);

    let y = ir.index(3);

    ir.schedule(&[&x, &y]);
    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(x.to_host_u32(), vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
    assert_eq!(y.to_host_u32(), vec![0, 1, 2]);
}
#[test]
fn load_gather() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.buffer_f32(&[1., 2., 3.]);

    ir.schedule(&[&x]);

    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(x.to_host_f32(), vec![1., 2., 3.]);
}
#[test]
fn eval_scatter() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.buffer_u32(&[0, 0, 0, 0]);
    let c = ir.literal_u32(1);
    let x = x.add(&c); // x: [1, 1, 1, 1]

    let i = ir.index(3);
    let c = ir.literal_u32(1);
    let i = i.add(&c); // i: [1, 2, 3]

    let y = ir.literal_u32(2);

    y.scatter(&x, &i, None); // x: [1, 2, 2, 2]

    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(x.to_host_u32(), vec![1, 2, 2, 2]);
}
#[test]
fn scatter_twice() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.buffer_u32(&[0, 0, 0, 0]);

    let i = ir.index(3);
    let c = ir.literal_u32(1);
    let i = i.add(&c);

    let y = ir.literal_u32(2);

    y.scatter(&x, &i, None);

    let i = ir.index(2);

    let y = ir.literal_u32(3);

    y.scatter(&x, &i, None);

    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(x.to_host_u32(), vec![3, 3, 2, 2]);
}
#[test]
fn scatter_twice_add() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.buffer_u32(&[0, 0, 0, 0]);

    let i = ir.index(3);
    let c = ir.literal_u32(1);
    let i = i.add(&c);

    let y = ir.literal_u32(2);

    y.scatter(&x, &i, None);

    let i = ir.index(2);

    let y = ir.literal_u32(3);

    y.scatter(&x, &i, None);

    let c = ir.literal_u32(1);
    let x = x.add(&c);

    ir.schedule(&[&x]);

    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(x.to_host_u32(), vec![4, 4, 3, 3]);
}
#[test]
fn scatter_reduce() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.buffer_u32(&[0, 0, 0, 0]);

    let i = ir.buffer_u32(&[0, 0, 0]);

    let y = ir.literal_u32(1);

    y.scatter_reduce(&x, &i, None, ReduceOp::Add);

    ir.schedule(&[&x]);

    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(x.to_host_u32(), vec![3, 0, 0, 0]);
}
#[test]
fn tex_lookup() {
    let ir = Trace::default();
    ir.set_backend("cuda");

    let x = ir.buffer_f32(&[0.5]);
    let y = ir.buffer_f32(&[0.5]);

    let data = ir.buffer_f32(&[1.; 400]);

    let tex = data.to_texture(&[10, 10]);

    let res = tex.tex_lookup(&[&x, &y]);

    let r = res[0].clone();

    r.schedule();

    let mut jit = Jit::default();
    jit.eval(&mut ir.borrow_mut());

    insta::assert_snapshot!(jit.kernel_debug());

    assert_eq!(r.to_host_f32(), vec![1.]);
}
