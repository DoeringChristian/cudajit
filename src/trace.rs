use std::fmt::Debug;
use std::ops::Deref;
use std::sync::Arc;

use bytemuck::cast_slice;
use once_cell::sync::Lazy;
use parking_lot::{MappedMutexGuard, Mutex, MutexGuard};
use slotmap::{DefaultKey, SlotMap};
use smallvec::smallvec;

use crate::backend::cuda::CUDABackend;
use crate::backend::Backend;
use crate::jit::{self, Jit};
use crate::var::{Op, Var, VarId, VarInfo, VarType};

// We have one global Intermediate Representation that tracks all operations.
// However, Other Intermediate representations can also be constructed.
pub static IR: Lazy<Trace> = Lazy::new(|| Trace::default());

///
/// A wrapper arrund an Intermediate Representation.
///
#[derive(Clone, Debug, Default)]
pub struct Trace(Arc<Mutex<Internal>>);
impl Deref for Trace {
    type Target = Arc<Mutex<Internal>>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Trace {
    fn push_var(&self, mut v: Var) -> VarRef {
        for dep in v.deps.iter() {
            self.lock().inc_rc(*dep);
        }
        v.rc = 1;
        let id = VarId(self.lock().vars.insert(v));
        VarRef::steal(&self, id)
    }
    fn var_info(&self, refs: &[&VarRef]) -> VarInfo {
        let ty = refs.first().unwrap().var().ty.clone(); // TODO: Fix (first non void)

        let size = refs
            .iter()
            .map(|id| id.var().size)
            .reduce(|s0, s1| s0.max(s1))
            .unwrap()
            .clone();
        VarInfo { ty, size }
    }
    fn push_var_op(&self, op: Op, deps: &[&VarRef], ty: VarType, size: usize) -> VarRef {
        let deps = deps
            .iter()
            .map(|r| {
                assert!(Arc::ptr_eq(&self.0, &r.ir));
                r.id()
            })
            .collect();
        let ret = self.push_var(Var {
            op,
            deps,
            ty,
            // param_ty: ParamType::None,
            buffer: None,
            size,
            rc: 0,
            literal: 0,
        });
        ret
    }
    pub fn set_backend(&self, backend: impl AsRef<str>) {
        if self.lock().backend.is_some() {
            return;
        }
        let backend = backend.as_ref();
        if backend == "cuda" {
            self.lock().backend = Some(Box::new(CUDABackend::new()));
        }
    }
    pub fn schedule(&self, refs: &[&VarRef]) {
        for r in refs {
            assert!(Arc::ptr_eq(&r.ir, &self.0));
            self.lock().schedule(&[r.id()])
        }
    }
}
impl Trace {
    // Constatns:
    pub fn const_f32(&self, val: f32) -> VarRef {
        self.push_var(Var {
            op: Op::Literal,
            deps: smallvec![],
            ty: VarType::F32,
            literal: bytemuck::cast::<_, u32>(val) as _,
            ..Default::default()
        })
    }
    pub fn const_u32(&self, val: u32) -> VarRef {
        self.push_var(Var {
            op: Op::Literal,
            deps: smallvec![],
            ty: VarType::U32,
            literal: bytemuck::cast::<_, u32>(val) as _,
            ..Default::default()
        })
    }
    pub fn const_bool(&self, val: bool) -> VarRef {
        self.push_var(Var {
            op: Op::Literal,
            deps: smallvec![],
            ty: VarType::Bool,
            literal: bytemuck::cast::<_, u8>(val) as _,
            ..Default::default()
        })
    }
    // Buffer initializers:
    pub fn buffer_f32(&self, slice: &[f32]) -> VarRef {
        let buffer = Some(
            self.lock()
                .backend
                .as_ref()
                .unwrap()
                .buffer_from_slice(cast_slice(slice)),
        );
        self.push_var(Var {
            // param_ty: ParamType::Input,
            buffer,
            size: slice.len(),
            ty: VarType::F32,
            op: Op::Data,
            ..Default::default()
        })
    }
    pub fn buffer_u32(&self, slice: &[u32]) -> VarRef {
        let buffer = Some(
            self.lock()
                .backend
                .as_ref()
                .unwrap()
                .buffer_from_slice(cast_slice(slice)),
        );
        self.push_var(Var {
            // param_ty: ParamType::Input,
            buffer,
            size: slice.len(),
            ty: VarType::U32,
            op: Op::Data,
            ..Default::default()
        })
    }
    // Special operations:
    pub fn index(&self, size: usize) -> VarRef {
        let v = self.push_var(Var {
            op: Op::Idx,
            deps: smallvec![],
            ty: VarType::U32,
            size,
            ..Default::default()
        });
        v
    }
}

#[derive(Default, Debug)]
pub struct Internal {
    vars: SlotMap<DefaultKey, Var>,
    pub backend: Option<Box<dyn Backend>>,
    pub functions: Vec<VarId>,
    pub scheduled: Vec<VarId>,
}

impl Internal {
    pub fn var(&self, id: VarId) -> &Var {
        &self.vars[id.0]
    }
    pub fn var_mut(&mut self, id: VarId) -> &mut Var {
        &mut self.vars[id.0]
    }
    pub fn get_var(&mut self, id: VarId) -> Option<&Var> {
        self.vars.get(id.0)
    }
    pub fn inc_rc(&mut self, id: VarId) {
        self.var_mut(id).rc += 1;
    }
    pub fn dec_rc(&mut self, id: VarId) {
        let var = self.var_mut(id);
        var.rc -= 1;
        if var.rc == 0 {
            for dep in var.deps.clone() {
                self.dec_rc(dep);
            }
            self.vars.remove(id.0);
        }
    }
    pub fn schedule(&mut self, ids: &[VarId]) {
        for id in ids {
            self.inc_rc(*id);
            self.scheduled.push(*id);
        }
    }
}

///
/// Reference to a variable.
///
#[derive(Debug)]
pub struct VarRef {
    id: VarId,
    ir: Trace,
}

impl VarRef {
    pub fn id(&self) -> VarId {
        self.id
    }
    pub fn borrow(trace: &Trace, id: VarId) -> Self {
        trace.lock().inc_rc(id);

        Self {
            id,
            ir: trace.clone(),
        }
    }
    pub fn steal(trace: &Trace, id: VarId) -> Self {
        Self {
            id,
            ir: trace.clone(),
        }
    }
    pub fn var(&self) -> MappedMutexGuard<Var> {
        MutexGuard::map(self.ir.lock(), |ir| &mut ir.vars[self.id().0])
    }
    pub fn is_data(&self) -> bool {
        let var = self.var();
        if var.buffer.is_some() {
            assert_eq!(var.op, Op::Data);
            true
        } else {
            assert_ne!(var.op, Op::Data);
            false
        }
    }
    pub fn schedule(&self) {
        self.ir.lock().schedule(&[self.id()])
    }
    // To Host functions:
    pub fn to_vec_f32(&self) -> Vec<f32> {
        let var = self.var();
        assert_eq!(var.ty, VarType::F32);
        let v = var.buffer.as_ref().unwrap().as_vec();
        Vec::from(cast_slice(&v))
    }
    pub fn to_vec_u32(&self) -> Vec<u32> {
        let var = self.var();
        assert_eq!(var.ty, VarType::U32);
        let v = var.buffer.as_ref().unwrap().as_vec();
        Vec::from(cast_slice(&v))
    }
    // Unarry operations:
    pub fn cast(&self, ty: VarType) -> VarRef {
        let v = self.var();
        self.ir.push_var(Var {
            op: Op::Cast,
            deps: smallvec![self.id()],
            ty,
            size: v.size,
            ..Default::default()
        })
    }
    // Binarry operations:
    pub fn add(&self, rhs: &VarRef) -> VarRef {
        let info = self.ir.var_info(&[self, &rhs]);
        self.ir
            .push_var_op(Op::Add, &[self, &rhs], info.ty, info.size)
    }
    pub fn mul(&self, rhs: &VarRef) -> VarRef {
        let info = self.ir.var_info(&[self, &rhs]);
        self.ir
            .push_var_op(Op::Mul, &[self, &rhs], info.ty, info.size)
    }
    pub fn and(&self, rhs: &VarRef) -> VarRef {
        assert!(Arc::ptr_eq(&self.ir, &rhs.ir));
        let info = self.ir.var_info(&[self, &rhs]);

        let ty_lhs = self.var().ty.clone();
        let ty_rhs = rhs.var().ty.clone();

        if info.size > 0 && ty_lhs != ty_rhs && !ty_rhs.is_bool() {
            panic!("Invalid operands!");
        }

        let ret = self
            .ir
            .push_var_op(Op::And, &[self, &rhs], info.ty, info.size);
        ret
    }
    // pub fn as_ptr(&self) -> Option<VarRef> {
    //     let ptr = self.var().buffer.as_ref().map(|b| b.as_ptr());
    //     // TODO: Eval var if needed.
    //     if let Some(ptr) = ptr {
    //         Some(self.ir.push_var(Var {
    //             op: Op::Literal,
    //             // param_ty: ParamType::Input,
    //             deps: smallvec![self.id()],
    //             ty: VarType::Ptr,
    //             literal: ptr,
    //             size: 1,
    //             stop_traversal: true,
    //             ..Default::default()
    //         }))
    //     } else {
    //         None
    //     }
    // }
    /// Reindex a variable with a new index and size.
    /// (Normally size is the size of the index)
    ///
    /// For now we construct a separate set of variables in the Ir.
    /// However, it should sometimes be possible to reuse the old ones.
    ///
    fn reindex(&self, new_idx: &VarRef, size: usize) -> Option<VarRef> {
        assert!(Arc::ptr_eq(&self.ir, &new_idx.ir));
        // let mut ir = self.ir.lock();

        let v = self.var();
        let is_literal = v.is_literal();
        let is_data = v.is_data();
        let v_deps = v.deps.clone();

        if is_data {
            return None;
        }

        drop(v);

        let mut deps = smallvec![];
        if !is_literal {
            for dep in v_deps {
                let dep = VarRef::borrow(&self.ir, dep);
                if let Some(dep) = dep.reindex(new_idx, size) {
                    deps.push(dep.id());
                } else {
                    return None;
                }
            }
        }

        let v = self.var();

        let op = v.op;
        let ty = v.ty.clone();
        // let param_ty = v.param_ty;
        let literal = v.literal;
        drop(v);

        if op == Op::Idx {
            // self.inc_rc(new_idx);
            return Some(new_idx.clone());
        } else {
            return Some(self.ir.push_var(Var {
                op,
                deps,
                ty,
                buffer: None,
                size,
                // param_ty,
                rc: 0,
                literal,
            }));
        }
    }
    ///
    /// For gather operations there are three ways to resolve them:
    ///
    /// If the source is a Pointer (i.e. VarType::Ptr) we can simply gather from that
    /// pointer.
    ///
    /// If the source is trivial (i.e. there are no dependencies on Input variablse
    /// ParamType::Input) we can
    /// reindex the variable.
    ///
    /// Finally, if the variable depends on Inputs we need to launch multiple Kernels (this
    /// is not yet implemented).
    ///
    pub fn gather(&self, index: &VarRef, mask: Option<&VarRef>) -> VarRef {
        assert!(Arc::ptr_eq(&self.ir, &index.ir));

        let size = index.var().size;
        let ty = self.var().ty.clone();

        let mask: VarRef = mask
            .map(|m| {
                assert!(Arc::ptr_eq(&self.ir, &m.ir));
                m.clone()
            })
            .unwrap_or(self.ir.const_bool(true));

        // let res = self.as_ptr();

        if self.var().buffer.is_some() {
            let ret = self.ir.push_var(Var {
                op: Op::Gather,
                deps: smallvec![self.id(), index.id(), mask.id()],
                ty,
                size,
                ..Default::default()
            });
            return ret;
        }

        let res = self.reindex(&index, size);

        if let Some(res) = res {
            let res = res.and(&mask);
            return res;
        }

        self.schedule(); // TODO: instead of evaluation, use dependency scheduling
        let mut jit = Jit::default();
        jit.eval(&mut self.ir.lock());

        if self.var().buffer.is_some() {
            let ret = self.ir.push_var(Var {
                op: Op::Gather,
                deps: smallvec![self.id(), index.id(), mask.id()],
                ty,
                size,
                ..Default::default()
            });
            return ret;
        }

        unimplemented!();
    }
}

impl Clone for VarRef {
    fn clone(&self) -> Self {
        self.ir.lock().inc_rc(self.id);
        Self {
            ir: self.ir.clone(),
            id: self.id,
        }
    }
}

impl Drop for VarRef {
    fn drop(&mut self) {
        self.ir.lock().dec_rc(self.id);
    }
}
