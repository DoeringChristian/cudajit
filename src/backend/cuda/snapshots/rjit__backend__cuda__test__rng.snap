---
source: src/backend/cuda/test.rs
expression: jit.kernel_debug()
---
===============================================
Kernel 14233316816652473877:

.version 8.0
.target sm_86
.address_size 64

.entry cujit(
	.param .align 8 .b8 params[32]) {

	.reg.b8   %b <10>; .reg.b16 %w<10>; .reg.b32 %r<10>;
	.reg.b64  %rd<10>; .reg.f32 %f<10>; .reg.f64 %d<10>;
	.reg.pred %p <10>;

	mov.u32 %r0, %ctaid.x;
	mov.u32 %r1, %ntid.x;
	mov.u32 %r2, %tid.x;
	mad.lo.u32 %r0, %r0, %r1, %r2; // r0 <- Index

	// Index Conditional (jump to done if Index >= Size).
	ld.param.u32 %r2, [params]; // r2 <- params[0] (Size)
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r0 >= r2
	@%p0 bra done; // if p0 => done
	
	mov.u32 %r3, %nctaid.x; // r3 <- nctaid.x
	mul.lo.u32 %r1, %r3, %r1; // r1 <- r3 * r1
	
body: // sm_86

	// [0]: ScheduleVar { op: Data, deps: [], ty: U32, param_ty: Input, reg: 7, buf: Some(0), tex: None, accel: None, literal: 0, size: 2 } =>
	ld.param.u64 %rd0, [params+8];
	mad.wide.u32 %rd0, %r0, 4, %rd0;
	ld.global.cs.u32 %r7, [%rd0];

	// [1]: ScheduleVar { op: Literal, deps: [], ty: U32, param_ty: None, reg: 8, buf: None, tex: None, accel: None, literal: 1065353216, size: 1 } =>
	mov.b32 %r8, 0x3f800000;


	// [2]: ScheduleVar { op: Or, deps: [SVarId(0), SVarId(1)], ty: U32, param_ty: None, reg: 6, buf: None, tex: None, accel: None, literal: 0, size: 2 } =>
	or.b32 %r6, %r7, %r8;

	// [3]: ScheduleVar { op: Literal, deps: [], ty: U32, param_ty: None, reg: 9, buf: None, tex: None, accel: None, literal: 9, size: 1 } =>
	mov.b32 %r9, 0x9;


	// [4]: ScheduleVar { op: Shr, deps: [SVarId(2), SVarId(3)], ty: U32, param_ty: None, reg: 5, buf: None, tex: None, accel: None, literal: 0, size: 2 } =>
	shr.b32 %r5, %r6, %r9;

	// [5]: ScheduleVar { op: Bitcast, deps: [SVarId(4)], ty: F32, param_ty: Output, reg: 4, buf: Some(1), tex: None, accel: None, literal: 0, size: 2 } =>
    mov.b32 %f4, %r5;

	// Store:
	ld.param.u64 %rd0, [params + 16]; // rd0 <- params[offset]
	mad.wide.u32 %rd0, %r0, 4, %rd0; // rd0 <- Index * ty.size() + params[offset]
	st.global.cs.f32 [%rd0], %f4; // (Index * ty.size() + params[offset])[Index] <- var

	//End of Kernel:

	add.u32 %r0, %r0, %r1; // r0 <- r0 + r1
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r1 >= r2
	@!%p0 bra body; // if p0 => body


done:

	ret;
}

